### PR TITLE
Auto-gen the 'mm' Op in the new lazy tensor core.

### PR DESCRIPTION
**Pull request by nkreeger**
This looks to be a straight-forward Op to port. The existing tooling auto-gens and everything seems to be working.

Are there any existing unit tests I should add/update here? I have a sample test file that I've verified with locally:

```py
from lazy_tensor_core import debug
import torch
import lazy_tensor_core
import lazy_tensor_core.debug.metrics as metrics

lazy_tensor_core._LAZYC._ltc_init_ts_backend()

torch.manual_seed(42)

device = 'lazy'
dtype = torch.float32

x = torch.randn(3, 3, device=device, dtype=dtype)
y = torch.randn(3, 3, device=device, dtype=dtype)

print(torch.mm(x, y))
print(metrics.metrics_report())
```

The diff from running the same code with the legacy op vs. the auto-generated one looks like this:
```diff
▶ diff -u legacy-output.txt new-output.txt
--- legacy-output.txt   2021-09-23 07:10:12.431555000 -0500
+++ new-output.txt      2021-09-23 07:11:00.441555000 -0500
@@ -3,21 +3,21 @@
         [ 0.3243,  2.8696,  2.7954]], device='lazy:0')
 Metric: DeviceLockWait
   TotalSamples: 21
-  Accumulator: 010.400us
-  ValueRate: 251.506us / second
-  Rate: 507.849 / second
-  Percentiles: 1%=000.300us; 5%=000.400us; 10%=000.400us; 20%=000.400us; 50%=000.500us; 80%=000.600us; 90%=000.700us; 95%=000.700us; 99%=000.700us
+  Accumulator: 011.800us
+  ValueRate: 279.624us / second
+  Rate: 497.635 / second
+  Percentiles: 1%=000.400us; 5%=000.400us; 10%=000.400us; 20%=000.500us; 50%=000.500us; 80%=000.600us; 90%=000.700us; 95%=000.900us; 99%=001.000us
 Metric: IrValueTensorToDataHandle
   TotalSamples: 4
-  Accumulator: 016.500us
-  ValueRate: 483.657us / second
-  Rate: 117.25 / second
-  Percentiles: 1%=001.400us; 5%=001.400us; 10%=001.400us; 20%=001.400us; 50%=003.400us; 80%=010.000us; 90%=010.000us; 95%=010.000us; 99%=010.000us
+  Accumulator: 016.200us
+  ValueRate: 472.753us / second
+  Rate: 116.729 / second
+  Percentiles: 1%=001.600us; 5%=001.600us; 10%=001.600us; 20%=001.600us; 50%=003.700us; 80%=009.200us; 90%=009.200us; 95%=009.200us; 99%=009.200us
 Metric: TensorsGraphSize
   TotalSamples: 21
   Accumulator: 81.00
-  ValueRate: 2126.05 / second
-  Rate: 551.197 / second
+  ValueRate: 2078.44 / second
+  Rate: 538.855 / second
   Percentiles: 1%=1.00; 5%=3.00; 10%=3.00; 20%=3.00; 50%=3.00; 80%=5.00; 90%=5.00; 95%=6.00; 99%=9.00
 Counter: CachedCompile
   Value: 1
@@ -57,8 +57,6 @@
   Value: 2
 Counter: lazy::lt
   Value: 1
-Counter: lazy::mm
-  Value: 1
 Counter: lazy::mul
   Value: 1
 Counter: lazy::ne
```

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp [hunk 1]
```diff
@@ -635,20 +635,6 @@ at::Tensor LazyNativeFunctions::max_pool3d(
       self, kernel_size, stride, padding, dilation, ceil_mode);
 }
 
-at::Tensor LazyNativeFunctions::mm(const at::Tensor& self,
-                                   const at::Tensor& mat2) {
-  LTC_FN_COUNTER("lazy::");
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:640]: ahh i was wondering why the 'counter' disappeared from the metrics output diff you pasted.  I guess I missed putting a counter into the codegenned ops.

File: lazy_tensor_core/lazy_tensor_core/csrc/tensor_methods.cpp [hunk 2]
```diff
@@ -1746,10 +1746,6 @@ void LazyTensor::min_out(LazyTensor& min, LazyTensor& min_indices,
   min.SetIrValue(ir::Value(node, 0));
   min_indices.SetIrValue(ir::Value(node, 1));
 }
-LazyTensor LazyTensor::mm(const LazyTensor& input, const LazyTensor& weight) {
-  return input.CreateFrom(
-      ir::ops::Dot(input.GetIrValue(), weight.GetIrValue()));
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/tensor_methods.cpp:1751]: presumably someone else is still using Dot class, but if not, we'd also want to remove that.  I didn't check..

**Comment by wconstab** [comment]
looks good, thanks @nkreeger 

RE testing, you should run test/cpp/build/test_ptltc (but CI also should have run this).  All the tests should pass, although, if you don't have Lapack installed you can ignore 9 test failures that mention Lapack missing.

**Comment by wconstab** [comment]
I'm actually not certain if the current tests directly cover the op you added.

This is the output from your run (fyi) - in the ` pytorch_linux_xenial_cuda11_1_cudnn8_py3_gcc7_test1 ` job
https://circleci.com/api/v1.1/project/github/pytorch/pytorch/16167496/output/107/0?file=true&allocation-id=614c7f7f5cd33572d19d9455-0-build%2F75CF9A17

This is the test src
lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp

there are a bunch of matmul tests, but they use more complex operators and i'm not sure if those are decomposing to mm or not in this case. An easy way to check is just to take the old non-codegen code and break it and see if the test fails.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp (hunk 1 above)
Comment (nkreeger) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:640]: Added this macro call into the code gen thingy.

File: lazy_tensor_core/lazy_tensor_core/csrc/tensor_methods.cpp (hunk 2 above)
Comment (nkreeger) [reply lazy_tensor_core/lazy_tensor_core/csrc/tensor_methods.cpp:1751]: Yeah looks like at least two ops use it in tensor_methods.cpp. Added a TODO - lemme know if more stuff needs to happen here.

**Comment by nkreeger** [comment]
New diff with counter thing:

```diff
▶ diff -u legacy-output.txt new-output.txt
--- legacy-output.txt   2021-09-23 07:10:12.431555000 -0500
+++ new-output.txt      2021-09-23 14:09:39.396254700 -0500
@@ -3,21 +3,21 @@
         [ 0.3243,  2.8696,  2.7954]], device='lazy:0')
 Metric: DeviceLockWait
   TotalSamples: 21
-  Accumulator: 010.400us
-  ValueRate: 251.506us / second
-  Rate: 507.849 / second
-  Percentiles: 1%=000.300us; 5%=000.400us; 10%=000.400us; 20%=000.400us; 50%=000.500us; 80%=000.600us; 90%=000.700us; 95%=000.700us; 99%=000.700us
+  Accumulator: 012.700us
+  ValueRate: 151.183us / second
+  Rate: 249.988 / second
+  Percentiles: 1%=000.400us; 5%=000.400us; 10%=000.400us; 20%=000.400us; 50%=000.500us; 80%=000.700us; 90%=000.800us; 95%=000.900us; 99%=001.800us
 Metric: IrValueTensorToDataHandle
   TotalSamples: 4
-  Accumulator: 016.500us
-  ValueRate: 483.657us / second
-  Rate: 117.25 / second
-  Percentiles: 1%=001.400us; 5%=001.400us; 10%=001.400us; 20%=001.400us; 50%=003.400us; 80%=010.000us; 90%=010.000us; 95%=010.000us; 99%=010.000us
+  Accumulator: 017.600us
+  ValueRate: 496.904us / second
+  Rate: 112.933 / second
+  Percentiles: 1%=001.400us; 5%=001.400us; 10%=001.400us; 20%=001.400us; 50%=005.400us; 80%=009.000us; 90%=009.000us; 95%=009.000us; 99%=009.000us
 Metric: TensorsGraphSize
   TotalSamples: 21
   Accumulator: 81.00
-  ValueRate: 2126.05 / second
-  Rate: 551.197 / second
+  ValueRate: 1002.89 / second
+  Rate: 260.008 / second
   Percentiles: 1%=1.00; 5%=3.00; 10%=3.00; 20%=3.00; 50%=3.00; 80%=5.00; 90%=5.00; 95%=6.00; 99%=9.00
 Counter: CachedCompile
   Value: 1
```

File: tools/codegen/dest/lazy_ir.py [hunk 3]
```diff
@@ -168,6 +168,7 @@ def gen_lazy_nativefunc_definition(f: NativeFunction, backend_index: BackendInde
 
     return [f"""\
 {sig.decl(name=f"{class_method_name}::{schema.aten_name}")} {{
+    LTC_FN_COUNTER("lazy::");
```
Comment (wconstab) [inline tools/codegen/dest/lazy_ir.py:171]: awesome! meant to actually point you to this code, but thanks for finding it.

**Review by wconstab — APPROVED** [review]

**Pull request merged**